tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package parser
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/config"
+)
+
+type operation int
+
+const (
+	orOp       = iota // OR = |
+	andOp             // AND = &
+	xorOp             // XOR = ^
+	lshiftOp          // LSHIFT = <<
+	rshiftOp          // RSHIFT = >>
+	onescompOp        // ONESCOMP = ~
+	timesOp           // TIMES = *
+	landOp            // LAND = &&
+	lorOp             // LOR = ||
+	lequalOp          // LEQUAL = ==
+	negOp             // MINUS -x
+	plusOp            // A + B
+	minusOp           // A - B
+
+	exprTypeName    string = "Expression Type"
+	groupTypeName   string = "Group Type"
+	bufferTypeName  string = "Buffer Type"
+	pointerTypeName string = "Pointer Type"
+)
+
+// TraceTree struct contains intermediate representation of trace
+// If a trace is multiprocess it constructs a trace for each type
+type TraceTree struct {
+	TraceMap map[int64]*Trace
+	Ptree    map[int64][]int64
+	RootPid  int64
+	Filename string
+}
+
+// NewTraceTree initializes a TraceTree
+func NewTraceTree() (tree *TraceTree) {
+	tree = &TraceTree{
+		TraceMap: make(map[int64]*Trace),
+		Ptree:    make(map[int64][]int64),
+		RootPid:  -1,
+	}
+	return
+}
+
+func (tree *TraceTree) contains(pid int64) bool {
+	if _, ok := tree.TraceMap[pid]; ok {
+		return true
+	}
+	return false
+}
+
+func (tree *TraceTree) add(call *Syscall) *Syscall {
+	if tree.RootPid < 0 {
+		tree.RootPid = call.Pid
+	}
+	if !call.Resumed {
+		if !tree.contains(call.Pid) {
+			tree.TraceMap[call.Pid] = newTrace()
+			tree.Ptree[call.Pid] = make([]int64, 0)
+		}
+	}
+	c := tree.TraceMap[call.Pid].add(call)
+	if c.CallName == "clone" && !c.Paused {
+		tree.Ptree[c.Pid] = append(tree.Ptree[c.Pid], c.Ret)
+	}
+	return c
+}
+
+// Trace is just a list of system calls
+type Trace struct {
+	Calls []*Syscall
+}
+
+func newTrace() *Trace {
+	return &Trace{Calls: make([]*Syscall, 0)}
+}
+
+func (trace *Trace) add(call *Syscall) (ret *Syscall) {
+	if call.Resumed {
+		lastCall := trace.Calls[len(trace.Calls)-1]
+		lastCall.Args = append(lastCall.Args, call.Args...)
+		lastCall.Paused = false
+		lastCall.Ret = call.Ret
+		ret = lastCall
+	} else {
+		trace.Calls = append(trace.Calls, call)
+		ret = call
+	}
+	return
+}
+
+// IrType is the intermediate representation of the strace output
+// Every argument of a system call should be represented in an intermediate type
+type IrType interface {
+	Name() string
```

Comment:
This Name seems to be used only in 2 places:
1. in Call.String, and there I think we actually want to use arg.String instead of arg.Name, because we want values, right?
2. Error message in inetAddr. There we can just print arg using `%#v` which will print actual type and fields, which much more useful for debugging than just "Field type" or "Expression type".
If we do that, then we can remove Name altogether.

Reply:
There are also few uses in proggen, but I looked at few and they are also error messages, so `%#v` will be more useful there too.

Reply:
And if we remove it, we can also remove expressionCommon and everything related to it.